crates/accelerate/src/remove_diagonal_gates_before_measure.rs

```diff
@@ -0,0 +1,103 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+/// Remove diagonal gates (including diagonal 2Q gates) before a measurement.
+use hashbrown::HashSet;
+use pyo3::prelude::*;
+
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::operations::StandardGate;
+
+/// Run the RemoveDiagonalGatesBeforeMeasure pass on `dag`.
+/// Args:
+///     dag (DAGCircuit): the DAG to be optimized.
+/// Returns:
+///     DAGCircuit: the optimized DAG.
+#[pyfunction]
+#[pyo3(name = "remove_diagonal_gates_before_measure")]
+fn run_remove_diagonal_before_measure(dag: &mut DAGCircuit) -> PyResult<()> {
+    static DIAGONAL_1Q_GATES: [StandardGate; 8] = [
+        StandardGate::RZGate,
+        StandardGate::ZGate,
+        StandardGate::TGate,
+        StandardGate::SGate,
+        StandardGate::TdgGate,
+        StandardGate::SdgGate,
+        StandardGate::U1Gate,
+        StandardGate::PhaseGate,
+    ];
+    static DIAGONAL_2Q_GATES: [StandardGate; 7] = [
+        StandardGate::CZGate,
+        StandardGate::CRZGate,
+        StandardGate::CU1Gate,
+        StandardGate::RZZGate,
+        StandardGate::CPhaseGate,
+        StandardGate::CSGate,
+        StandardGate::CSdgGate,
+    ];
+
+    let mut nodes_to_remove = HashSet::new();
```

Comment:
I'm not quite sure here, but I think this here would even be better as Vector, since we never have to search in this so we can skip calculating the hash 🤔

Reply:
Indeed, we can change it to a Vec, but at the end we need to convert it to a HashSet before removing the nodes. It slightly improves the performance.

| n_qubits | depth | time is sec  | 
|----------|--------|--------------|
| 5 | 1024 | 0.00 |  
| 14 | 1024 | 0.02 |  
| 20 | 1024 | 0.04 | 
| 50 | 1024 |  0.31 |  
| 100 | 1024 | 1.48 |